Initial implementation ir:user

mailwl opened the pull request:
code from https://github.com/citra-emu/citra/issues/1292
Circle Pad Pro still not supported.

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-21T11:57:05Z: https://github.com/citra-emu/citra/compare/0d086616d1af239b8e41b246bea3c1ef85fc907f...mailwl:899f9ed5ecb93debaa740b091bc1545548298e5d

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-21T12:03:10Z: https://github.com/citra-emu/citra/compare/0d086616d1af239b8e41b246bea3c1ef85fc907f...mailwl:d7ca85782b6f6e6c63c571a63b6f770b0c4b4282

lioncash commented on src/core/hle/service/ir/ir.cpp:
```diff
@@ -26,6 +27,50 @@ void GetHandles(Service::Interface* self) {
+    conn_status_event->Signal();
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+
+    LOG_INFO(Service_IR, "(STUBBED) called: param1 = %d", param1);
```
%u

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-21T12:51:17Z: https://github.com/citra-emu/citra/compare/0d086616d1af239b8e41b246bea3c1ef85fc907f...mailwl:d008f98c76aa7704504e384bfcf2b454abf1198e

lioncash commented:
Squash your changes

mailwl commented:
i did this, but nothing changes, is it enough?


wl@DESKTOP MINGW64 /c/projects/citra (branch-ir)
$ git rebase -i HEAD~3
Successfully rebased and updated refs/heads/branch-ir.

$ git rebase -i upstream/master
Successfully rebased and updated refs/heads/branch-ir.

$ git push origin branch-ir --force
Everything up-to-date

lioncash commented:
> is it enough?

No, because you did it incorrectly somehow

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-21T14:16:53Z: https://github.com/citra-emu/citra/compare/0d086616d1af239b8e41b246bea3c1ef85fc907f...mailwl:645028998d2d682aeed2ea6cf3f2c0b316d160e5

ealibaba commented:
@mailwl  hi，i am a new developer, i have problems on android_server project, can you help me? my email is 2415966803@qq.com,thx

bunnei commented on src/core/hle/service/ir/ir.h:
```diff
@@ -20,6 +20,53 @@ namespace IR {
 void GetHandles(Interface* self);
 
+/**
+* IR::InitializeIrNopShared service function
```
All of these lines that start with '*' should have a space in front of them

bunnei commented on src/core/hle/service/ir/ir.cpp:
```diff
@@ -26,6 +27,50 @@ void GetHandles(Service::Interface* self) {
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+
+    LOG_INFO(Service_IR, "(STUBBED) called");
```
We're not 100% consistent, but if you're logging a "stubbed" service routine call, it should be LOG_WARNING

bunnei commented on src/core/hle/service/ir/ir.cpp:
```diff
@@ -26,6 +27,50 @@ void GetHandles(Service::Interface* self) {
+    conn_status_event->Signal();
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+
+    LOG_INFO(Service_IR, "(STUBBED) called: param1 = %u", param1);
```
Should be "param1=%u" to be consistent (spaces removed)

bunnei replied:
Also, what's the point of printing this unknown param? Above (for InitializeIrNopShared) you have a known parameter (baud rate) but are not printing it there?

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-26T07:49:21Z: https://github.com/citra-emu/citra/compare/0d086616d1af239b8e41b246bea3c1ef85fc907f...mailwl:085451dd0a3e04aa99ae7b81ab509407663103b6

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-26T07:58:26Z: https://github.com/citra-emu/citra/compare/2b00bdec1f83be82acb65bb316086858134e5766...mailwl:804968609bcb4f66531c8c29332913a3f8c2ec7c

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-26T08:13:30Z: https://github.com/citra-emu/citra/compare/2b00bdec1f83be82acb65bb316086858134e5766...mailwl:b52cbf4d15301dd9a90969c41fd2d34258070d58

bunnei commented on src/core/hle/service/ir/ir.cpp:
```diff
@@ -35,15 +95,19 @@ void Init() {
 
     using Kernel::MemoryPermission;
     shared_memory = SharedMemory::Create(0x1000, Kernel::MemoryPermission::ReadWrite,
```
Is this unused? From looking at what you've implemented, it seems like the game would allocate the shared memory region and then pass it to the IR service (e.g. the `transfer_shared_memory` you added).

mailwl replied:
it's from ir:rst namespace, for example GetHandles doesn't call in game Dragon Quest VIII

bunnei commented on src/core/hle/service/ir/ir.cpp:
```diff
@@ -26,6 +28,64 @@ void GetHandles(Service::Interface* self) {
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+
+    LOG_WARNING(Service_IR, "(STUBBED) called, transfer_buff_size = %d, recv_buff_size = %d, "
```
You missed one of my comments (whitespace nitpick when logging). This should be:
<pre>
LOG_WARNING(Service_IR, "(STUBBED) called, transfer_buff_size=%d, recv_buff_size=%d, "
            "unknown1=%d, send_buff_size=%d, unknown2=%d, baud_rate=%u",
            transfer_buff_size, recv_buff_size, unknown1, send_buff_size, unknown2, baud_rate);
</pre>
Also, IMO you can add the `handle` to that list as well, since that's usually useful when debugging.

bunnei commented:
Once you've addressed my last 2 comments, feel free to squash this down to 1 commit. Thanks.

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-26T15:34:26Z: https://github.com/citra-emu/citra/compare/c28a48aa023b1617dfc94897d677f4f731f66db8...mailwl:3c195219d6f8339f7309bbc7275f686ce3075a82

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-26T15:49:06Z: https://github.com/citra-emu/citra/compare/c28a48aa023b1617dfc94897d677f4f731f66db8...mailwl:0783d0e4aee3cf229f622a9a3734e11368569cd8

neobrain commented:
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-26T15:53:21Z: https://github.com/citra-emu/citra/compare/c28a48aa023b1617dfc94897d677f4f731f66db8...mailwl:2e13b7f3ca9e7ecf0b0b6c1da65453e46ee6728f

bunnei commented:
LGTM

bunnei commented:
Fixes #1292

purpasmart96 commented on src/core/hle/service/ir/ir.cpp:
```diff
@@ -35,15 +95,19 @@ void Init() {
     using Kernel::MemoryPermission;
     shared_memory = SharedMemory::Create(0x1000, Kernel::MemoryPermission::ReadWrite,
-            Kernel::MemoryPermission::ReadWrite, "IR:SharedMemory");
+                                         Kernel::MemoryPermission::ReadWrite, "IR:SharedMemory");
```
lol what happened here
